sim/pkg/model/replicas_active.go

```diff
@@ -41,30 +42,47 @@ func (ras *replicasActiveStock) Count() uint64 {
 }
 
 func (ras *replicasActiveStock) EntitiesInStock() []*simulator.Entity {
-	return ras.delegate.EntitiesInStock()
+	return ras.replicas
 }
 
-func (ras *replicasActiveStock) Remove() simulator.Entity {
-	entity := ras.delegate.Remove()
-	if entity == nil {
+func (ras *replicasActiveStock) Remove(entity *simulator.Entity) simulator.Entity {
+	removedEntity := ras.delegate.Remove(entity)
+	if removedEntity == nil {
 		return nil
 	}
 
-	replica := entity.(Replica)
+	replica := removedEntity.(ReplicaEntity)
 	replica.Deactivate()
 
-	return entity
+	//support replicas array updated
+	ras.deleteReplica(entity)
+	return removedEntity
 }
 
 func (ras *replicasActiveStock) Add(entity simulator.Entity) error {
-	replica := entity.(Replica)
+	replica := entity.(ReplicaEntity)
 	replica.Activate()
+	ras.replicas = append(ras.replicas, &entity)
 	return ras.delegate.Add(entity)
 }
 
+func (ras *replicasActiveStock) deleteReplica(replica *simulator.Entity) {
+	removed := false
+	for i := 0; i < len(ras.replicas); i++ {
+		if ras.replicas[i] == replica {
+			ras.replicas[i] = ras.replicas[len(ras.replicas)-1]
+			removed = true
+		}
+	}
+	if removed {
+		ras.replicas = ras.replicas[:len(ras.replicas)-1]
+	}
+}
+
 func NewReplicasActiveStock(env simulator.Environment) ReplicasActiveStock {
 	return &replicasActiveStock{
 		env:      env,
-		delegate: simulator.NewThroughStock("ReplicasActive", "Replica"),
+		delegate: simulator.NewHeterogenousThroughStock("ReplicasActive", "Replica"),
```

Comment:
So you are creating a heterogeneous delegate stock, which should have constant time add and delete.  However you are still maintaining a slice of replicas.  Which means you still get O(n) time deleting (see the for loop above).

Why do you maintain two stocks for replicas (delegate and replicas)?

Reply:
Deleted a slice of replicas